File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example2/Handlers/VirusScannerHandler.cs
namespace ChainOfResponsibilityDesignPattern.Example2.Handlers;

public class VirusScannerHandler : BaseFileHandler
{
    // Simulated virus signatures
    private readonly List<string> _virusSignatures = new() { "MALWARE", "VIRUS", "TROJAN" };

    public override void Handle(FileUploadRequest request)
    {
        // Simulate virus scanning by checking file content
        string contentAsString = System.Text.Encoding.UTF8.GetString(request.FileContent);
        
        bool virusDetected = _virusSignatures.Any(signature => 
            contentAsString.Contains(signature, StringComparison.OrdinalIgnoreCase));

        if (virusDetected)
        {
            request.IsValid = false;
            request.ValidationMessages.Add("Virus detected! File rejected.");
            return; // Stop the chain
        }
        
        // Pass to next handler
        base.Handle(request);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example2/src/Handlers/VirusScannerHandler.ts
import { FileUploadRequest } from "../FileUploadRequest";
import { BaseFileHandler } from "../BaseFileHandler";

const VIRUS_SIGNATURES = ["MALWARE", "VIRUS", "TROJAN"];

export class VirusScannerHandler extends BaseFileHandler {
  handle(request: FileUploadRequest): void {
    const content = request.fileContent.toString("utf8");
    const virusDetected = VIRUS_SIGNATURES.some((sig) =>
      content.toUpperCase().includes(sig)
    );
    if (virusDetected) {
      request.isValid = false;
      request.validationMessages.push("Virus detected! File rejected.");
      return;
    }
    super.handle(request);
  }
}
